stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/event/handler/internal/EventHandlerSequential.kt

```diff
@@ -0,0 +1,727 @@
+                        channelId = channelState.channelId
+                    )
+                    channelLogic.handleEvent(userPresenceChanged)
+                }
+        }
+
+        // only afterwards forward to the queryRepo since it borrows some data from the channel
+        // queryRepo mainly monitors for the notification added to channel event
+        logicRegistry.getActiveQueryChannelsLogic().forEach { channelsLogic ->
+            channelsLogic.handleEvents(sortedEvents)
+        }
+    }
+
+    private suspend fun updateOfflineStorage(batchEvent: BatchEvent) {
+        logger.v { "[updateOfflineStorage] batchEvent.size: ${batchEvent.size}" }
+        val currentUserId: UserId = currentUserId.get() ?: error("no current userId provided")
+
+        val events = batchEvent.sortedEvents
+        val batchBuilder = EventBatchUpdate.Builder()
+        batchBuilder.addToFetchChannels(events.filterIsInstance<CidEvent>().map { it.cid })
+
+        val users: List<User> = events.filterIsInstance<UserEvent>().map { it.user } +
+            events.filterIsInstance<HasOwnUser>().map { it.me }
+
+        batchBuilder.addUsers(users)
+
+        // step 1. see which data we need to retrieve from offline storage
+
+        val messageIds = events.extractMessageIds()
+        batchBuilder.addToFetchMessages(messageIds)
+
+        // actually fetch the data
+        val batch = batchBuilder.build(repos, currentUserId)
+
+        // step 2. second pass through the events, make a list of what we need to update
+        for (event in events) {
+            when (event) {
+                is ConnectedEvent -> if (batchEvent.isFromSocketConnection) {
+                    event.me.id mustBe currentUserId
+                    repos.insertCurrentUser(event.me)
+                }
+                // keep the data in Room updated based on the various events..
+                // note that many of these events should also update user information
+                is NewMessageEvent -> {
+                    event.message.enrichWithCid(event.cid)
+                    event.message.enrichWithOwnReactions(batch, currentUserId, event.user)
+                    batch.addMessageData(event.cid, event.message, isNewMessage = true)
+                    repos.selectChannelWithoutMessages(event.cid)?.copy(hidden = false)
+                        ?.let(batch::addChannel)
+                }
+                is MessageDeletedEvent -> {
+                    event.message.enrichWithCid(event.cid)
+                    event.message.enrichWithOwnReactions(batch, currentUserId, event.user)
+                    batch.addMessageData(event.cid, event.message)
+                }
+                is MessageUpdatedEvent -> {
+                    event.message.enrichWithCid(event.cid)
+                    event.message.enrichWithOwnReactions(batch, currentUserId, event.user)
+                    batch.addMessageData(event.cid, event.message)
```

Comment:
I think some code here is duplicated in `EventHandlerImpl`. It would be nice to merge both of them because we will have to remember to always fix in both places if we have to change the code here. It can be challenging to keep both implementations the same as time passes.

Reply:
Hey @leandroBorgesFerreira,

That's because of the agreement to keep `EventHandlerImpl` to avoid breaking anything and introduce a new `EventHandler` to replace the existing one if no issues are found.